test/org/javarosa/core/model/ItemsetBindingTest.java

```diff
@@ -0,0 +1,147 @@
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+
+import static org.javarosa.core.test.AnswerDataMatchers.stringAnswer;
+import static org.javarosa.core.test.SelectChoiceMatchers.choice;
+
+public class ItemsetBindingTest {
+    private Scenario scenario;
+
+    @Before
+    public void setUp() {
+        scenario = Scenario.init("three-level-cascading-select.xml");
+    }
+
+    @Test
+    public void dependentLevelsInBlankInstance_ShouldHaveNoChoices() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+    }
+
+    @Test
+    public void selectingValueAtLevel1_ShouldFilterChoicesAtLevel2() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+
+        scenario.answer("/data/level1", "b");
+
+        assertThat(scenario.choicesOf("/data/level2"), containsInAnyOrder(
+            choice("ba"),
+            choice("bb"),
+            choice("bc")));
+    }
+
+    @Test
+    public void selectingValuesAtLevels1And2_ShouldFilterChoicesAtLevel3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "b");
+        scenario.answer("/data/level2", "ba");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("baa"),
+            choice("bab")));
+    }
+
+    @Test
+    public void clearingValueAtLevel2_ShouldClearChoicesAtLevel3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "a");
+        scenario.answer("/data/level2", "aa");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("aaa"),
+            choice("aab")));
+        scenario.answer("/data/level2", "");
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+    }
+
+    @Test
+    public void clearingValueAtLevel1_ShouldClearChoicesAtLevels2And3() {
+        scenario.newInstance();
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        assertThat(scenario.choicesOf("/data/level3"), empty());
+
+        scenario.answer("/data/level1", "a");
+        scenario.answer("/data/level2", "aa");
+        assertThat(scenario.choicesOf("/data/level3"), containsInAnyOrder(
+            choice("aaa"),
+            choice("aab")));
+
+        scenario.answer("/data/level1", "");
+        assertThat(scenario.choicesOf("/data/level2"), empty());
+        // this next assertion is only true because the one before called populateDynamic choices
+        // TODO: make clearing out answers that are no longer available choices part of the form re-evaluation
```

Comment:
Does this have to do with the new way Collect can show many selects on the same page?

Reply:
Kind of. 😝 The underlying issue is what I described at https://github.com/opendatakit/javarosa/pull/437#discussion_r284443873 and that affects any client whether questions are on one screen or not. Ideally we could remove that `choicesOf` assertion and the following two assertions would pass. That's not the case right now because we need to force a re-computation of the second level's choices and `choicesOf` does that (as it should and as any client would do to display choices).

With the fix I'm proposing here, clients can get correct behavior *but only if they request the choices of all selects in a cascade in order*. With selects on the same screen, some questions might get removed when the screen is refreshed so the current implementation loops backwards over questions. So the slightly irritating thing is that I need another loop that runs forwards to recompute the choices. It's not a very big deal, though, it's the broader performance implication of having to re-compute the same choice lists over and over again that is a problem.

Reply:
Thanks for the clarification! I think I have the whole context now :)